pkg/services/export/service.go

```diff
@@ -0,0 +1,126 @@
+package export
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+type ExportService interface {
+	// List folder contents
+	HandleGetStatus(c *models.ReqContext) response.Response
+
+	// Read raw file contents out of the store
+	HandleRequestExport(c *models.ReqContext) response.Response
+}
+
+type StandardExport struct {
+	logger log.Logger
+	sql    *sqlstore.SQLStore
+	mutex  sync.Mutex
+
+	// updated with mutex
+	status ExportStatus
+	cfg    ExportConfig
+}
+
+func ProvideService(sql *sqlstore.SQLStore, features featuremgmt.FeatureToggles) ExportService {
+	if !features.IsEnabled(featuremgmt.FlagExport) {
+		return &StubExport{}
+	}
+
+	return &StandardExport{
+		sql:    sql,
+		logger: log.New("export"),
+	}
+}
+
+func (ex *StandardExport) HandleGetStatus(c *models.ReqContext) response.Response {
+	ex.mutex.Lock()
+	defer ex.mutex.Unlock()
+
+	if ex.status.Started > 0 {
+		return response.JSON(http.StatusOK, ex.status)
+	}
+
+	// The system is not currently running
+	return response.JSON(http.StatusOK, ExportStatus{
+		Running: false,
+		Changed: time.Now().UnixMilli(),
+	})
+}
+
+func (ex *StandardExport) HandleRequestExport(c *models.ReqContext) response.Response {
+	var cfg ExportConfig
+	err := json.NewDecoder(c.Req.Body).Decode(&cfg)
+	if err != nil {
+		return response.Error(http.StatusBadRequest, "unable to read config", err)
+	}
+
+	if cfg.Format != "git" {
+		return response.Error(http.StatusBadRequest, "only git format supported for now", err)
+	}
+
+	ex.mutex.Lock()
+	defer ex.mutex.Unlock()
+
+	if ex.status.Running {
+		return response.Error(http.StatusLocked, "export already running", nil)
+	}
+
+	ex.cfg = cfg
+	ex.status = ExportStatus{
+		Running: true,
+		Target:  "git export",
+		Started: time.Now().UnixMilli(),
+	}
+	go ex.doExport()
```

Comment:
I do not like that "doExport" is not passed the config, and it actually updates the cfg/status... but this structure has the mutex set config and unlock in this function

Any better ideas?

Reply:
Tried to come up with something on `export-service-job` branch - I extracted the job-running logic from `service.go` and added a `Job` interface - the export service now cares only about starting jobs & managing job instances